### PR TITLE
fix(checkpoint): support 64 GB RTX Spark training

### PR DESCRIPTION
**Pull request by akoumpa**
# What does this PR do ?

Fix NVBug 6674647 checkpoint loading on integrated-memory CUDA devices and add explicit 64 GB RTX Spark recipe variants.

# Changelog

- Prefault full-state safetensors only when the destination is an integrated CUDA device, while preserving mmap-backed/reclaimable checkpoint tensors.
- Stage one DCP mmap slice at a time before an integrated-GPU copy; CPU and discrete-GPU paths retain their existing behavior.
- Make `LlamaStateDictAdapter` implement the shared `StateDictAdapter` contract without model-specific checkpoint dispatch.
- Add dedicated `_rtx_spark` recipes for Llama 3.1 8B LoRA, Qwen3 8B LoRA, and Llama 3.3 70B QLoRA; existing generic and `_spark` recipes are unchanged.
- Encode the 64 GB UMA settings and regression matrix in focused tests and short critical-path comments.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?

Validation:

- `pytest tests/unit_tests/checkpoint/test_checkpointing.py tests/unit_tests/recipes/llm/test_rtx_spark_configs.py -q`: 214 passed, 5 skipped.
- Focused Ruff format and lint checks: passed.
- Discrete RTX 5880 Llama custom/HF output equivalence: 2 passed.
- RTX Spark (`NVIDIA JMJWOA-Generic-GPU`, SM12x), `nvcr.io/nvidia/nemo-automodel:26.08`, Docker `--memory=64g --memory-swap=64g`:
  - Ran `llama3_1_8b_squad_peft_rtx_spark.yaml` directly with only `max_steps=1` overridden.
  - Completed model load, forward, backward, optimizer step, validation, and checkpoint save.
  - Step 0: loss 0.2735, grad norm 20.5473, reported memory 15.28 GiB.
  - Exit 0; `OOMKilled=false`.
  - Artifact: `/home/nvidia/spark_playbook_runs/20260831T034717Z-existing-branch-rtx-yaml`.
- The custom Llama mmap path loads the 8B checkpoint in 1.02 s without OOM. SM12x cannot run TE packed backward (`ragged LSE` cuDNN limitation), so the end-to-end RTX recipe intentionally uses HF FlashAttention-2 plus activation checkpointing.

# Additional Information

Related to NVBug 6674647.

### CODE REVIEW

**Comment by akoumpa** [comment]
RTX Spark 64 GB UMA validation at head `29852f694e2c937cce7355e878e7530e900e7db0`:

- Container: `nvcr.io/nvidia/nemo-automodel:26.08`, `--memory=64g --memory-swap=64g`, 1 GPU.
- Llama 3.1 8B LoRA: packed size 128, steps 0-99, stable 15.39 GiB, final val loss 0.2874, checkpoint saved, exit 0, OOMKilled=false. Remote artifact: `/home/nvidia/spark_playbook_runs/20260831T055800Z-llama3_1_8b-packed100-disable-mmap`.
- Qwen3 8B LoRA: packed size 128, steps 0-99, stable 15.55 GiB, final val loss 0.7102, checkpoint saved, exit 0, OOMKilled=false. Remote artifact: `/home/nvidia/spark_playbook_runs/20260831T054700Z-qwen3_8b-packed100-disable-mmap`.
- `disable_mmap: true` reduced the cold Qwen force-HF load from >20 minutes without reaching step 0 to ~10 seconds to tokenizer/dataset setup.
- Recipe config tests: 7 passed.

The packed Llama 3.3 70B QLoRA run is still pending its 131.4 GiB gated checkpoint staging; I will append that result separately.

**Comment by akoumpa** [comment]
70B RTX Spark validation for head `111d4e3b9b29022226f6c256b6d5d22322370520` is passing now.

Setup:
- Remote RTX Spark host
- Container image: `nvcr.io/nvidia/nemo-automodel:26.08`
- Docker limits: `--memory=64g --memory-swap=64g --gpus all --ipc=host`
- Config: `llama_3_3_70b_instruct_squad_peft_qlora_rtx_spark.yaml`
- Overrides: `--step_scheduler.max_steps 100 --model.pretrained_model_name_or_path /workspace/model`
- Packed sequence active: `packed_sequence_size: 1024`; log: `THD-packing dataset with size: 1024`, `Total number of packs created: 21074`
- Model cache: local-backed 70B safetensors snapshot, 30 shards, `141107497872` bytes

Evidence:
- Container: `rtx-packed100-llama70b-bytes-20260831t084648`
- Run dir: `/localhome/swqa/cawang_n1x/dl-mobile-test-runner/automodel_rtx_spark_6674647/runs/20260831T084648Z-llama3_3_70b-packed100-stream-bytes`
- Docker inspect: `exit=0 oom=false`, started `2026-08-31T08:47:35Z`, finished `2026-08-31T10:17:18Z`
- Streaming load: shard 1 at `08:47:45`, shard 30 at `08:50:09`, load complete at `08:50:13` (`723 tensors loaded`)
- Training: `step 0` at `08:51:59`; `step 99` at `10:03:42`; progress bar reached `100/100`
- Sampled memory: Docker cgroup ~`43.0-44.4GiB / 61.6GiB`; train log steady at `40.48 GiB` after warmup
- Checkpoint: `checkpoints/epoch_0_step_99` completed; `.incomplete` removed; `adapter_model.safetensors` (~414MB) and optimizer distcp (~834MB) written; final validation log `loss 0.1359`

Local focused tests also pass:
`python -m pytest -q tests/unit_tests/_transformers/test_model_init.py::TestStreamLoadBnbWeights tests/unit_tests/_transformers/test_model_init.py::TestBnbModulesToNotConvert tests/unit_tests/_transformers/test_model_init.py::TestStreamingBnbSupported tests/unit_tests/recipes/llm/test_rtx_spark_configs.py`

Result: 70B QLoRA + packed sequences now trains for 100 steps on the 64GB unified-memory RTX Spark setup without OOM, and checkpoint save completes.

File: nemo_automodel/_transformers/model_init.py [hunk 1]
```diff
@@ -809,42 +842,26 @@ def _stream_load_bnb_weights(model, model_dir, device, torch_dtype):
...
-                else:
-                    target_dtype = torch_dtype if torch_dtype is not None else tensor.dtype
-                    materialized = tensor.to(device=device, dtype=target_dtype)
-                    del tensor
-                    if isinstance(old_param, torch.nn.Parameter):
-                        mod._parameters[attr] = torch.nn.Parameter(materialized, requires_grad=old_param.requires_grad)
-                    else:
-                        mod._buffers[attr] = materialized
+            # On UMA systems, CUDA faults over file-backed checkpoint pages can
+            # dominate load time. Read only the active shard into anonymous
```
Comment (yuhezhang-ai) [inline nemo_automodel/_transformers/model_init.py]: [P1 — important / merge-blocking] Keep non-mmap BnB loading bounded for unsharded checkpoints

With `disable_mmap=True`, `f.read()` materializes the entire current safetensors file before any tensor is quantized. The loader also explicitly supports the unsharded `model.safetensors` fallback, so a consolidated 70B BF16 checkpoint can exceed the 64 GB Spark memory budget immediately. This also contradicts the function’s documented one-tensor peak-memory bound.

Please preserve tensor-bounded staging—for example, use `safe_open()` and clone one tensor at a time into anonymous host memory before passing it to `consume_tensor()`—or reject oversized unsharded checkpoints with an actionable sharding requirement.

File: nemo_automodel/components/checkpoint/_backports/hf_storage.py [hunk 2]
```diff
@@ -321,18 +329,30 @@ def read_data(self, plan: LoadPlan, planner: LoadPlanner) -> Future[None]:
...
                         assert target_tensor.size() == tensor.size(), (
                             f"req {req.storage_index} mismatch sizes {target_tensor.size()} vs {tensor.size()}"
                         )
 
+                        # On integrated GPUs, copying a file-backed mmap directly to CUDA can
+                        # migrate it page-by-page and take minutes per tensor. Stage only this
+                        # requested slice in anonymous host memory; unlike the full-state loader,
+                        # peak host memory is bounded by one tensor. Discrete GPUs keep the
+                        # zero-copy mmap source used to avoid host OOM on very large checkpoints.
+                        if file_backed and _is_integrated_cuda_device(target_tensor.device):
```
Comment (yuhezhang-ai) [inline nemo_automodel/components/checkpoint/_backports/hf_storage.py:353]: [P2 — important coverage gap] Exercise the integrated-GPU DCP staging branch

The new integrated-device branch is not exercised by the added tests. The current tests cover device classification and the separate full-state prefault loader, but not the actual `read_data()` sequence of aligned mmap access, narrowing, anonymous staging, destination copy, and commit.

Please add a focused reader test using a local safetensors file, force the integrated-device predicate, request a narrowed slice, and verify the exact committed values. Since the checked-in Spark recipes use either `force_hf` or BnB streaming, a small RTX Spark DCP smoke test would also be useful if this branch remains in scope.

File: nemo_automodel/_transformers/model_init.py [hunk 3]
```diff
@@ -800,6 +799,40 @@ def _stream_load_bnb_weights(model, model_dir, device, torch_dtype):
     loaded_keys: set[str] = set()
     device = torch.device(device) if not isinstance(device, torch.device) else device
 
+    def consume_tensor(key: str, tensor: torch.Tensor) -> None:
```
Comment (yuhezhang-ai) [inline nemo_automodel/_transformers/model_init.py:804]: [Documentation — non-blocking] Document the tensor contract for `consume_tensor`

This new tensor-handling function needs the repository-required Google-style tensor contract. Please document that `tensor` may have arbitrary rank and axis order determined by its FQN, along with its dtype/device handling and ownership behavior: the value may be converted and quantized for `Params4bit`, or installed as a parameter/buffer after device/dtype conversion. This is documentation cleanup rather than a runtime correctness concern.

File: nemo_automodel/components/models/llama/state_dict_adapter.py [hunk 4]
```diff
@@ -75,3 +80,9 @@ def to_hf(
         if exclude_key_regex is not None:
             return {k: v for k, v in state_dict.items() if not re.search(exclude_key_regex, k)}
         return dict(state_dict)
+
+    def convert_single_tensor_to_hf(self, fqn: str, tensor: Any, **kwargs) -> list[tuple[str, Any]]:
```
Comment (yuhezhang-ai) [inline nemo_automodel/components/models/llama/state_dict_adapter.py:84]: [Documentation — non-blocking] Document the passthrough tensor layout and aliasing

Please add the repository-required Google-style `Args` and `Returns` contract for `convert_single_tensor_to_hf`. The tensor may have arbitrary rank and axis order determined by its FQN, and the returned tuple contains the same tensor object without copying or transformation, so the output aliases the input. This is documentation cleanup rather than a runtime correctness concern.

**Comment by akoumpa** [comment]
Update after Yuhe's review comments, current head `c9646c68e6` (`c9646c68e659125034f3f3aff542aaad7dfd4af6`).

Addressed locally:
- P1: `disable_mmap=True` BnB loading no longer materializes a whole safetensors shard with `f.read()`. It now iterates with `safe_open()` and clones only the target tensor before `consume_tensor()`.
- P2: added a focused DCP reader test that forces the integrated-device staging branch, reads a narrowed safetensors slice, and verifies exact committed values.
- Docs: added Google-style tensor-contract docs for `consume_tensor` and LLaMA `convert_single_tensor_to_hf`.

Local validation:
- `python -m pytest -q tests/unit_tests/_transformers/test_model_init.py::TestStreamLoadBnbWeights tests/unit_tests/_transformers/test_model_init.py::TestBnbModulesToNotConvert tests/unit_tests/_transformers/test_model_init.py::TestStreamingBnbSupported tests/unit_tests/checkpoint/_backports/test_hf_storage_reader.py tests/unit_tests/recipes/llm/test_rtx_spark_configs.py` -> `21 passed`
- `python -m ruff check ...` -> passed
- `python -m ruff format --check ...` -> passed
- `git diff --check` -> passed

Pending verification:
- RTX Spark 26.08 / 64GB remote smoke has not been rerun yet. `10.24.167.136:22` is reachable, but the supplied account paths currently reject auth. I will rerun the remote smoke once access is restored.

File: nemo_automodel/_transformers/model_init.py (hunk 1 above)
Comment (akoumpa) [reply nemo_automodel/_transformers/model_init.py]: Addressed in `c9646c68e6`. The `disable_mmap=True` path no longer uses `f.read()` / full-shard deserialization. It now uses `safe_open()` and clones one target tensor at a time before `consume_tensor()`, so unused tensors are not staged. Local targeted tests pass; RTX Spark rerun is still pending because the remote host currently rejects the supplied SSH auth.

File: nemo_automodel/components/checkpoint/_backports/hf_storage.py (hunk 2 above)
Comment (akoumpa) [reply nemo_automodel/components/checkpoint/_backports/hf_storage.py:353]: Addressed in `c9646c68e6`. Added `tests/unit_tests/checkpoint/_backports/test_hf_storage_reader.py`, which creates a local safetensors file, forces the integrated-device predicate, reads a narrowed slice, and checks the target plus committed tensor values exactly. Local targeted tests pass; RTX Spark DCP smoke is pending remote access restoration.

File: nemo_automodel/_transformers/model_init.py (hunk 3 above)
Comment (akoumpa) [reply nemo_automodel/_transformers/model_init.py:804]: Addressed in `c9646c68e6`. Added the Google-style tensor contract for `consume_tensor`, including FQN semantics, tensor ownership, and cast/quantize/move/install behavior.

File: nemo_automodel/components/models/llama/state_dict_adapter.py (hunk 4 above)
Comment (akoumpa) [reply nemo_automodel/components/models/llama/state_dict_adapter.py:84]: Addressed in `c9646c68e6`. Added the Google-style passthrough contract for `convert_single_tensor_to_hf`, including key-driven layout semantics and explicit input aliasing in the returned tuple.

**Comment by akoumpa** [comment]
Validation update for `3983be273e2263b6475cb22b5d9fb4e25eed5c36`:

- CI was retriggered with `/ok to test 3983be273e2263b6475cb22b5d9fb4e25eed5c36`; latest `gh pr checks` snapshot shows all checks passing.
- Local focused validation passed:
  - `pytest` checkpoint/model-init/step-scheduler/RTX-Spark recipe tests: `64 passed`
  - `ruff check` and `ruff format --check` on touched files passed
- RTX Spark 64 GB UMA validation passed on the 26.08 container with Docker constrained to `--memory=64g --memory-swap=64g --gpus all`.
  - Exact remote git head: `3983be273e2263b6475cb22b5d9fb4e25eed5c36`, clean tree
  - Recipe/config path: `examples/llm_finetune/llama3_3/llama_3_3_70b_instruct_squad_peft_qlora_rtx_spark.yaml`
  - Runtime overrides: `max_steps=100`, `ckpt_every_steps=100`, `save_consolidated=false`
  - Confirmed config: `validate_on_checkpoint=false`, `val_every_steps=null`, `packed_sequence_size=1024`
  - Log evidence: streaming BnB load completed (`723 tensors loaded`), packed sequences created (`21074` packs), train completed through step 99, then saved checkpoint to `/workspace/run/checkpoints/epoch_0_step_99`
  - Container result: `exit=0`, `OOMKilled=false`
  - Checkpoint artifact: `1.2G`, including `model/adapter_model.safetensors` and `optim/__0_0.distcp`
  - GPU process list was empty after container cleanup

Remote log retained at:
`/home/nvidia/automodel_pr3766_remote_20260901/logs/70b_exact3983_no_val_100step_20260901T222900Z.log`

**Review by yuhezhang-ai — APPROVED** [review]
